plugins/hls-pragmas-plugin/test/Main.hs

```diff
@@ -74,7 +75,10 @@ codeActionTest testComment fp actions =
     _ <- waitForDiagnosticsFrom doc
     cas <- map fromAction <$> getAllCodeActions doc
     mapM_ (\(action, contains) -> go action contains cas) actions
-    executeCodeAction $ head cas
+    action <- case cas of
```

Comment:
Instead of explicit `case` everywhere, wouldn't you instead write a function, such as:

```haskell
unsafeHeadWithNoteM :: (HasCallStack, MonadIO m) => String -> [a] -> m a
unsafeHeadWithNoteM message l = case l of
   [] -> liftIO $ assertFailure message
   (x:_) -> pure a
```

Or something similar?

Reply:
Similar function could indeed be used, but I'm not sure where to put it, in order for it to be usable by different test suites. In this particular test suite it could only be used once as the other case requires failure also on lists of length 2+

Reply:
I actually did not realized that it was code in the test suite. I don't really care about "code quality", "deduplication" and "exception safety" in test suite, so you can consider my suggestion as not important.

I'll let a project owner continue the discussion.

Reply:
No problem. I figured eliminating some partial function usages in tests can still be useful, as in this recent example:
https://github.com/haskell/haskell-language-server/pull/2639#issuecomment-1022727133

Reply:
For tests, I'm wondering if running tests with `-prof` enabled may not gives free stacktrace by default. It will be a bit slower, but with nicer error messages.

Note: at work, we use `base-noprelude` + a custom `Prelude.hs` which just reexport `Prelude` minors `head` and a few others, just rewrite `head` with `HasCallstack` constraint. We changed nothing to the codebase and immediatly got better stacktrace. This can be another option.

Reply:
We're also using base-noprelude at work, unfortunately that package hasn't been maintained for a long time (we had to fork it to be able to upgrade to ghc 8.10.7). Moreover switching to custom prelude approach would be much larger and more controversial change than simpler "let's just not use partial functions".

Reply:
I do agree. Good luck to progress on that patch, that's a cool merge request.

Reply:
I'm pro removing partial functions in tests still, but I don't think it's worth exerting too much effort on making it super tidy. We do have `hls-test-utils`, though.

Reply:
Indeed, hls-test-utils seems like a good place for that utility. I moved it there and reused in few more places.